### PR TITLE
rpc: rejected quoted decimal block-number strings; require 0x hex format

### PR DESCRIPTION
**Pull request by lupin012**
To fix rpc-compact debug_traceBlockByNumber/trace-block-invalid-number:
  {"jsonrpc":"2.0","id":1,"method":"debug_traceBlockByNumber","params":["3"]}
  should be rejected

### CODE REVIEW

File: rpc/types.go [hunk 1]
```diff
@@ -107,37 +107,44 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
...
+		case "safe":
+			*bn = SafeBlockNumber
+			return nil
+		case "finalized":
+			*bn = FinalizedBlockNumber
+			return nil
+		case "latestExecuted":
+			*bn = LatestExecutedBlockNumber
+			return nil
+		case "null":
```
Comment (yperbasis) [inline rpc/types.go]: Moving `case "null"` inside the quoted-string branch silently changes behavior for the bare JSON `null` literal. Previously the switch ran regardless of quoting, so a bare `null` (which `encoding/json` does route here, as the literal bytes `null`) resolved to `LatestBlockNumber`. Now it falls through to the bare-integer path → `ParseUint("null")` fails → `DecodeUint64("null")` fails → error, while quoted `"null"` still resolves. A client that passed `null` for an optional block param and used to get `latest` will now get `-32602`.

This side effect isn't covered by the ChangeLog (which only mentions quoted decimals). Please confirm it's intended; if so, document it, and either way add a test pinning the bare-`null` behavior so it can't regress silently again.

File: rpc/types.go [hunk 2]
```diff
@@ -107,37 +107,44 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
...
+			*bn = LatestBlockNumber
+			return nil
+		}
+		blckNum, err := hexutil.DecodeUint64(input)
+		if err != nil {
+			return err
+		}
+		if blckNum > math.MaxInt64 {
```
Comment (yperbasis) [inline rpc/types.go]: This `> math.MaxInt64` check + `*bn = BlockNumber(blckNum)` + `return nil` tail is now duplicated verbatim between the quoted-hex branch (here) and the bare-integer branch (lines 151-154), including the identical `errors.New("block number larger than int64")` string. Consider having both parse paths converge on a single tail (parse to `blckNum`, then one shared bound-check + assign) to drop the duplication.

File: rpc/types.go [hunk 3]
```diff
@@ -107,37 +107,44 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
...
+			return errors.New("block number larger than int64")
+		}
+		*bn = BlockNumber(blckNum)
 		return nil
 	}
-
-	// Try to parse it as a number
+	// Bare JSON integer: accept decimal or hex.
```
Comment (yperbasis) [inline rpc/types.go]: The comment is inaccurate, and the fallback it describes is dead: a bare JSON integer can never be hex — `0x3` isn't valid JSON, so via `json.Unmarshal` this branch only ever receives an unquoted decimal token. The `hexutil.DecodeUint64` fallback on line 146 is therefore unreachable for real JSON input.

Per the repo CLAUDE.md ("Default: no comment." / "Not a restatement of the code"), suggest dropping the comment and the unreachable fallback. If some direct (non-`json`) caller relies on the hex fallback, keep it but make the comment say that instead.

**Review by yperbasis — CHANGES_REQUESTED** [review]
Requesting changes on a few review items. The core change — rejecting quoted decimal block numbers and requiring `0x` hex — looks correct, and the tests + ChangeLog cover it well. Inline comments below for the bare-`null` behavior change, the duplicated bound-check tail, and the misleading comment + dead hex fallback on the bare-integer path.

One item that doesn't map to a changed line:

**Stale docstring (lines 100-105):** the `UnmarshalJSON` doc still says it supports "the block number" and lists only the named tags, with no mention that quoted numeric input must now be `0x` hex (quoted decimal is rejected). Since this is the user-visible contract being changed, please update the docstring to match.

File: rpc/types.go [hunk 4]
```diff
@@ -107,37 +107,44 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
...
-		return nil
-	case "finalized":
-		*bn = FinalizedBlockNumber
-		return nil
-	case "latestExecuted":
-		*bn = LatestExecutedBlockNumber
-		return nil
-	case "null":
-		*bn = LatestBlockNumber
+		switch input {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply rpc/types.go]: fixed

File: rpc/types.go (hunk 1 above)
Comment (lupin012) [reply rpc/types.go]: fixed: Restored with a single early-return that handles both null and "null" before the quoted/unquoted branch. Tests added

File: rpc/types.go (hunk 2 above)
Comment (lupin012) [reply rpc/types.go]: done

File: rpc/types.go (hunk 3 above)
Comment (lupin012) [reply rpc/types.go]: done

File: rpc/mcp/mcp_test.go [hunk 5]
```diff
@@ -74,7 +74,7 @@ func TestParseBlockNumber(t *testing.T) {
 		{
 			name:    "decimal number",
 			input:   "100",
-			wantErr: false,
+			wantErr: true,
```
Comment (yperbasis) [inline rpc/mcp/mcp_test.go:78]: **MCP: decimal block numbers now silently resolve to the earliest block**

Good to pin this in the test. The follow-up concern is on the production side. `parseBlockNumber` rejects quoted decimals now because `rpc/mcp/utils.go:25` wraps the input in quotes before `UnmarshalJSON`, and most MCP handlers discard the returned error, e.g.

```go
// rpc/mcp/mcp.go:552 — handleGetTransactionByBlockNumberAndIndex
blockNum, _ := parseBlockNumber(req.GetString("blockNumber", "latest"))
```

`parseBlockNumber("100")` now returns an error, so `blockNum` keeps its zero value `rpc.BlockNumber(0)`, which is `EarliestBlockNumber`. A request that used to read block 100 now silently reads the earliest block instead of erroring.

This applies to the other `, _ :=` call sites (e.g. lines 516, 632, 646, 656, 693, 732, 819, 862, 886, 933) and the `if bn, err := ...; err == nil` guards in `handleGetLogs` (589/594), which silently drop the bound. Handlers that check the error — e.g. `handleGetBlockByNumber` (line 459) — surface it correctly.

The error-swallowing predates this PR, but this change is what turns a previously-valid decimal input into a silent wrong-block read. Worth either propagating the error in those handlers (as `handleGetBlockByNumber` does) or documenting that MCP block-number arguments must be hex or a named tag.

File: rpc/types_test.go [hunk 6]
```diff
@@ -50,6 +50,8 @@ func TestBlockNumberJSONUnmarshal(t *testing.T) {
 		14: {`someString`, true, BlockNumber(0)},
 		15: {`""`, true, BlockNumber(0)},
 		16: {``, true, BlockNumber(0)},
+		17: {`"3"`, true, BlockNumber(0)},
+		18: {`"100"`, true, BlockNumber(0)},
 	}
```
[the remark this answers is not in the document]
Comment (lupin012) [reply rpc/types_test.go:57]: done

File: rpc/mcp/mcp_test.go (hunk 5 above)
Comment (lupin012) [reply rpc/mcp/mcp_test.go:78]: done

**Review by yperbasis — APPROVED** [review]
